### PR TITLE
✨ Melee unit ai

### LINKED ISSUE #34: Melee Unit AI

### PR DESCRIPTION
**Pull request by 429277**
A prototype for the Melee Unit AI has been added with pathfinding and basic attack animations.
Closes #34

### CODE REVIEW

**Review by NathanThus — CHANGES_REQUESTED** [review]
Please solve the merge conflict, then re-request a review.

File: Assets/Animations/Cube 1.controller
Comment (NathanThus) [inline Assets/Animations/Cube 1.controller:1]: What is Cube 1? It's not a very descriptive name.

File: Assets/Animations/MeleeUnit 1.controller
Comment (NathanThus) [inline Assets/Animations/MeleeUnit 1.controller:1]: This needs a better name. Is it a generic controller? Sword? Spear?

File: Assets/Animations/Melee_Unit_Attack.anim
Comment (NathanThus) [inline Assets/Animations/Melee_Unit_Attack.anim:1]: What unit?

File: Assets/Materials/Wall.mat
Comment (NathanThus) [inline Assets/Materials/Wall.mat:1]: Do you really need this?
How will this contribute to the project?

File: Assets/Scenes/MeleeUnitAI.unity
Comment (NathanThus) [inline Assets/Scenes/MeleeUnitAI.unity:1]: Is this a testing scene? Demo?

File: Assets/Scripts/Units/Melee Unit/MeleeUnit.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace RogueApeStudio.Crusader.Units.MeleeUnit
+{
+    public class MeleeUnit : MonoBehaviour
+    {
+        public UnitMovement LocalUnitMovement;
```
Comment (NathanThus) [inline Assets/Scripts/Units/Melee Unit/MeleeUnit.cs]: Does not conform to our code standard for fields.
- Public
- Lacks underscore prefix.
- PascalCase

File: Assets/Scripts/Units/Melee Unit/MeleeUnit.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        public Animator LocalAnimator;
+
+        public float StartAttackDistance = 2.0f;
+        [SerializeField] private float movementSpeed = 3.5f;
```
Comment (NathanThus) [inline Assets/Scripts/Units/Melee Unit/MeleeUnit.cs]: Lacks underscore prefix.

File: Assets/Scripts/Units/Melee Unit/MeleeUnit.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+        public IMeleeUnitState CurrentState;
+        public IMeleeUnitState ChaseState = new MeleeUnitChaseState();
+        public IMeleeUnitState AttackState = new MeleeUnitAttackState();   
+
+        private void Awake()
+        {
+            CurrentState = ChaseState;
+            LocalUnitMovement.SetStopDistance(StartAttackDistance - 0.2f);
```
Comment (NathanThus) [inline Assets/Scripts/Units/Melee Unit/MeleeUnit.cs]: No magic numbers please.

File: Assets/Scripts/Units/Melee Unit/MeleeUnitAttackState.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    public class MeleeUnitAttackState : IMeleeUnitState
+    {
+        public void EnterState(MeleeUnit meleeUnit)
+        {
+            meleeUnit.LocalAnimator.SetTrigger("Attack");
+        }
+
+        public void ExitState(MeleeUnit meleeUnit)
+        {
+
```
Comment (NathanThus) [inline Assets/Scripts/Units/Melee Unit/MeleeUnitAttackState.cs]: Empty method, why does it exist?

File: Assets/Scripts/Units/Melee Unit/MeleeUnitChaseState.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+
+namespace RogueApeStudio.Crusader.Units.MeleeUnit
+{
+    public class MeleeUnitChaseState : IMeleeUnitState
```
Comment (NathanThus) [inline Assets/Scripts/Units/Melee Unit/MeleeUnitChaseState.cs]: Too many empty functions.

File: Assets/Scripts/Units/PlayerTracker.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+
+namespace RogueApeStudio.Crusader.Units
+{
+    public class PlayerTracker : MonoBehaviour
+    {
+        public static PlayerTracker instance;
```
Comment (NathanThus) [inline Assets/Scripts/Units/PlayerTracker.cs]: No public fields.

**Review by NathanThus — CHANGES_REQUESTED** [review]
Please take another look at the code convention.

File: Assets/Animations/Cube 1.controller
Comment (NathanThus) [inline Assets/Animations/Cube 1.controller:1]: What is cube 1?

File: Assets/Scenes/MeleeUnitAI.unity
Comment (NathanThus) [reply Assets/Scenes/MeleeUnitAI.unity:1]: @429277 A bit clearer, but why is this scene included?
Do we need this for development, or was this a testing thing?

File: Assets/Materials/Wall.mat
Comment (NathanThus) [reply Assets/Materials/Wall.mat:1]: @429277 Still unanswered.

File: Assets/Animations/MeleeUnit 1.controller
Comment (NathanThus) [reply Assets/Animations/MeleeUnit 1.controller:1]: Unresolved.

File: Assets/Animations/Cube 1.controller
Comment (NathanThus) [reply Assets/Animations/Cube 1.controller:1]: Unresolved

File: Assets/Scripts/Units/Melee Unit/MeleeUnit.cs (hunk 1 above)
Comment (NathanThus) [reply Assets/Scripts/Units/Melee Unit/MeleeUnit.cs]: Unresolved.

File: Assets/Scripts/Units/Melee Unit/MeleeUnit.cs (hunk 2 above)
Comment (NathanThus) [reply Assets/Scripts/Units/Melee Unit/MeleeUnit.cs]: Unresolved.

File: Assets/Scripts/Units/Melee Unit/AxeUnitChaseState.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+
+namespace RogueApeStudio.Crusader.Units.MeleeUnit
+{
+    public class AxeUnitChaseState : IAxeUnitState
+    {
+        public void EnterState(AxeUnit AxeUnit)
```
Comment (NathanThus) [inline Assets/Scripts/Units/Melee Unit/AxeUnitChaseState.cs]: From a style point of view, make it an empty one liner.
Also, the arguments do not conform to the code convetion. They should be camelCase.

```csharp
public void EnterState(AxeUnit axeUnit) { }
```

**Review by NathanThus — CHANGES_REQUESTED** [review]
Please take another look at the requested changes.
Additionally, please revert this commit, as you should not merge `feature/melee-unit-ai` into `feature/melee-unit-ai`.

`8cc6266765928b76cd51d9eb53545713bb887549`

File: Assets/Art/PolygonAdventure/Materials/PolyAdventureMaterial_Water_01.mat [hunk 8]
```diff
@@ -36,6 +36,7 @@ Material:
     RenderType: Transparent
   disabledShaderPasses:
   - DepthOnly
+  - SHADOWCASTER
```
Comment (NathanThus) [inline Assets/Art/PolygonAdventure/Materials/PolyAdventureMaterial_Water_01.mat:39]: Why is this disabled?

File: Assets/Art/PolygonAdventure/Materials/PolyAdventureMaterial_Water_01.mat [hunk 9]
```diff
@@ -133,7 +134,7 @@ Material:
     - _ZWrite: 0
     m_Colors:
     - _BaseColor: {r: 0.6985294, g: 0.82535493, b: 1, a: 0.24705882}
-    - _Color: {r: 1, g: 1, b: 1, a: 1}
+    - _Color: {r: 0.6985294, g: 0.82535493, b: 1, a: 0.24705882}
```
Comment (NathanThus) [inline Assets/Art/PolygonAdventure/Materials/PolyAdventureMaterial_Water_01.mat:137]: Why did the colours need to change?

File: Assets/Scenes/Sandbox.unity
Comment (NathanThus) [inline Assets/Scenes/Sandbox.unity:1]: Not what I meant when I said "Create a sandbox folder".
You need to create a `Sandbox` folder in `Assets/Scenes` and drag your scene IN that folder.
So it becomes `Assets/Scenes/Sandbox/YourTestingSceneGoesHere.unity`

File: Assets/Scripts/Units/Axe Unit/AxeUnit.cs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+{
+    public class AxeUnit : MonoBehaviour
+    {
+        [SerializeField] private float _stopDistance = 1.8f;
+        [SerializeField] private float _movementSpeed = 3.5f;
+        [SerializeField] private float _startAttackDistance = 2.0f;
+        [SerializeField] private UnitMovement _localUnitMovement;
+        [SerializeField] private Animator _localAnimator;
+
+        public UnitMovement LocalUnitMovement => _localUnitMovement;
```
Comment (NathanThus) [inline Assets/Scripts/Units/Axe Unit/AxeUnit.cs:19]: Almost!
The order is fields, then properties.

File: Assets/Scripts/Units/UnitMovement.cs [hunk 11]
```diff
@@ -0,0 +1,24 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.AI;
+namespace RogueApeStudio.Crusader.Units
+{
+    public class UnitMovement : MonoBehaviour
+    {
+        [SerializeField] private NavMeshAgent _navMeshAgent;
```
Comment (NathanThus) [inline Assets/Scripts/Units/UnitMovement.cs:9]: Nitpick: spacing would be nice around here.

**Review by NathanThus — CHANGES_REQUESTED** [review]
Great improvements over the last review, there's just a few more things 😉

File: Assets/Art/PolygonAdventure/Materials/PolyAdventureMaterial_Water_01.mat (hunk 9 above)
Comment (429277) [reply Assets/Art/PolygonAdventure/Materials/PolyAdventureMaterial_Water_01.mat:137]: Uhhhhh I didn't even know I touched this

Comment (NathanThus) [reply Assets/Art/PolygonAdventure/Materials/PolyAdventureMaterial_Water_01.mat:137]: Then please revert it. If you need to, grab a copy from `develop`.

File: Assets/Scenes/Sandbox.unity
Comment (NathanThus) [reply Assets/Scenes/Sandbox.unity:1]: Unresolved.

**Review by NathanThus — CHANGES_REQUESTED** [review]
Last comments still standing.
Still, almost done :)

**Review by NathanThus — APPROVED** [review]
Nice work!

**Comment by NathanThus** [comment]
You still want to add an assembly definition to the `Axe Unit` folder, but the code is good.

**Pull request merged**